pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,20 +242,198 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
+				return err
+			}
+		}
+		return c.syncMCFlows()
+	}
+	if err := syncFn(); err == nil {
+		c.queue.Forget(key)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", key)
 	}
 	return true
 }
 
+// syncWireGuard reconciles WireGuard configurations in following way:
+//  1. If the current Node is the Multi-cluster Gateway Node, controller will try to initialize corresponding WireGuard
+//     configuration and route on the host, then add all existing Gateway Nodes in other member clusters as WireGuard peers.
+//  2. If the current Node is not Multi-cluster Gateway Node, controller will try to clean up WireGuard configurations.
+//
+// Note: MCDefaultRouteController runs only one worker to process Gateway and ClusterInfoImport. So we do not need
+// any synchronization mechanism.
+func (c *MCDefaultRouteController) syncWireGuard() error {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		return err
+	}
+
+	amIGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+	if c.wireGuardClient != nil && (!amIGateway || !c.wireGuardInitialized) {
+		if err := c.cleanUpWireGuard(); err != nil {
+			return err
+		}
+		c.wireGuardInitialized = false
+	}
+	if !amIGateway {
+		return nil
+	}
+	if !c.wireGuardInitialized {
+		var initErr error
+		defer func() {
+			if initErr != nil {
+				if err := c.cleanUpWireGuard(); err != nil {
+					klog.ErrorS(err, "Failed to clean up WireGuard")
+				}
+				c.wireGuardInitialized = false
+			}
+		}()
+		if initErr = c.initializeWireGuard(gateway); initErr != nil {
+			return initErr
+		}
+		c.wireGuardInitialized = true
+	}
+	ciImports, err := c.ciImportLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	desiredCIImports := sets.NewString()
+	var updateErr error
+	for _, ciImport := range ciImports {
+		desiredCIImports.Insert(ciImport.Name)
+		if err = c.updateSingleWireGuardPeer(ciImport); err != nil {
+			klog.V(2).ErrorS(err, "Failed to update WireGuard peer", "ClusterInfoImport", klog.KObj(ciImport))
+			updateErr = err
+		}
+	}
+	if updateErr != nil {
+		return updateErr
+	}
+
+	// Check cache and existing ClusterInfoImports, clean up routes and WireGuard peers of the
+	// removed ClusterInfoImports.
+	for ciName, ciImport := range c.installedCIImports {
+		if desiredCIImports.Has(ciName) {
+			continue
+		}
+		if err := c.removeWireGuardRouteAndPeer(ciImport); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MCDefaultRouteController) removeWireGuardRouteAndPeer(ciImport *mcv1alpha1.ClusterInfoImport) error {
+	remoteGatewayIP, _, _ := net.ParseCIDR(ciImport.Spec.ServiceCIDR)
+	dstCIDR := net.IPNet{IP: remoteGatewayIP, Mask: net.CIDRMask(32, 32)}
+	if err := c.routeClient.DeleteRouteForLink(&dstCIDR, c.wireGuardConfig.LinkIndex); err != nil {
```

Comment:
Can `DeleteRouteForLink` handle empty `dstCIDR` correctly if `ciImport.Spec.ServiceCIDR` is not a valid CIDR? I suppose you should just return nil when the `net.ParseCIDR(ciImport.Spec.ServiceCIDR)` return error instead of moving to next step directly.

Reply:
The `ciImport.Spec.ServiceCIDR` is got from `installedCIImport`, which checked when installed. So the ServiceCIDR may not be invalid. That's why I ignored the error.

Reply:
Got it, thanks.